clients/pkg/promtail/client/batch.go

```diff
@@ -23,37 +36,74 @@ type batch struct {
 	streams   map[string]*logproto.Stream
 	bytes     int
 	createdAt time.Time
+	wal       WAL
 }
 
-func newBatch(entries ...api.Entry) *batch {
+func newBatch(wal WAL) *batch {
 	b := &batch{
 		streams:   map[string]*logproto.Stream{},
 		bytes:     0,
 		createdAt: time.Now(),
+		wal:       wal,
 	}
+	return b
+}
 
-	// Add entries to the batch
-	for _, entry := range entries {
-		b.add(entry)
+// replay adds an entry without writing it to the log, to be used
+// when replaying WAL segments at startup
+func (b *batch) replay(entry api.Entry) {
+	labelsString := labelsMapToString(entry.Labels, ReservedLabelTenantID)
+	if stream, ok := b.streams[labelsString]; ok {
+		stream.Entries = append(stream.Entries, entry.Entry)
+		return
+	}
+	// Add the entry as a new stream
+	b.streams[labelsString] = &logproto.Stream{
+		Labels:  labelsString,
+		Entries: []logproto.Entry{entry.Entry},
 	}
-
-	return b
 }
 
 // add an entry to the batch
 func (b *batch) add(entry api.Entry) {
 	b.bytes += len(entry.Line)
+	walRecord.RefEntries = walRecord.RefEntries[:0]
+	walRecord.Series = walRecord.Series[:0]
+	hashBuffer = hashBuffer[:0]
+
+	// todo: log the error
+	defer func() {
+		err := b.wal.Log(walRecord)
+		if err != nil {
+			fmt.Println("err: ", err)
+		}
+	}()
+
+	var fp uint64
+	lbs := labels.FromMap(util.ModelLabelSetToMap(entry.Labels))
+	sort.Sort(lbs)
+	fp, hashBuffer = lbs.HashWithoutLabels(hashBuffer, []string(nil)...)
```

Comment:
Not sure the hashbuffer can be used safely with globals as multiple clients could collide in usage.

Reply:
We weren't actually using the hash for anything so I've removed it. If we do end up needing it we can use a `sync.Pool`